### ISSUE TITLE
Fail to replicate DiMP FPS in the paper (57 FPS)

### ISSUE THREAD

**Issue by shen338** [visionml/pytracking#231]
Hi, 

Thank you for this amazing work! 

I just ran into some problems when benchmarking tracker FPS. I am using Tesla P100 GPU, which should at the same level as GTX 1080 or higher. But the FPS is less than 20 with 0.06 second processing time per image. 

I followed the install.md to install pytracking and download dimp18.pth. The script I am using is run_video.py from ./pytracking/pytracking. All the config remain unchanged. I am using torch 1.7.1 and CUDA 10.2. 

Could you have any clue why this is happening? Thank you! 

Add more details here: 

Overall average processing time per image: ~0.06 second

Major time-consuming functions:
Line 306 in dimp.py: im_patches, patch_coords = sample_patch_multiscale(.....) takes ~0.015 second
Line 310 in dimp.py: backbone_feat = self.net.extract_backbone(im_patches) takes ~0.015 second
Line 128 in dimp.py: self.refine_target_box(backbone_feat, sample_pos[scale_ind,:], sample_scales[scale_ind], scale_ind, update_scale_flag) takes ~0.025 second

**Comment by martin-danelljan** [comment]
Hi. Thanks for your appreciation.

There are a number of factors that could be the cause:

* The 57 FPS is for dimp18. dimp50 was around 40 or so if I remember correctly.
* We have noticed significant performance differences with different PyTorch versions. Specifically, I think we used 1.1 or 1.0 back then. And we saw that newer versions were significantly slower, and we havent understood why really. Although we havent done a real timing for 1.6 and 1.7, so things could have improved.
* Although I agree that Tesla P100 and 1080ti should be similar since its the same architecture, there could be differences. I havent tried.
* CPU speed is also important for some parts. For example `sample_patch_multiscale` is on CPU. To me it is strange that this operation takes so much time for you. There are probably some optimizations to be done there.

**Comment by shen338** [comment]
Thank you for the thorough answer! 

I also noticed different versions of Torch produce very different FPS. May have to try 1.1 ~ 1.4 to benchmark this. For the sample_batch_multiscale, I will take a look and update my optimizations here. 

Happy new year!

**Comment by shen338** [comment]
@martin-danelljan Hi Martin, a quick question, could I use direct bbox regression instead of ATOM IoUNet to refine the target bounding box. This IoUNet bbox refine takes half of the processing time. 

`def refine_target_box(self, backbone_feat, sample_pos, sample_scale, scale_ind, update_scale = True):
        """Run the ATOM IoUNet to refine the target bounding box."""

        if hasattr(self.net.bb_regressor, 'predict_bb'):
            return self.direct_box_regression(backbone_feat, sample_pos, sample_scale, scale_ind, update_scale)

        # Initial box for refinement
        init_box = self.get_iounet_box(self.pos, self.target_sz, sample_pos, sample_scale)`

There is already a bb_regressor in DiMP code.

**Comment by martin-danelljan** [comment]
Hi. Yes there is some support for direct box regression. We have actually tried many types of direct BB regression (architectures and losses) but never been able to make it close to as good as the PrDiMP/SuperDiMP regressor. The models we have tried are not available in this repo (due to their inferior performance). But it should not be so difficult to implement if you want to try.

**Comment by shen338** [comment]
Good to know! My project has a lower requirement for localization accuracy, considering bounding box over 0.3 IOU as a successful tracking. I might need a bounding box regression to further improve the speed. 

I am facing another problem with the initialization part. When running the tracker on the live video (30FPS), the long tracker initialization time (>2 seconds) is problematic. I tried to increase the learning rate and reduce initialization iterations to speed it up, but it still takes more than 1 second. 

Do you have suggestions for this? Or any other reference materials for me to speed up tracker initialization? I was thinking a meta-learned pretrained filter should be a good choice, which can be adapted to classify any target object quickly enough. It may also give a lift to tracker performance as well.

**Comment by martin-danelljan** [comment]
Hi. This is more of a implementation thing. Also some parameters can increase the speed.
* First make sure that the network weights are already loaded before you start tracking (this takes quite some time).
* You could try to reduce the number of initial optimization iterations.
* The best thing otherwise is to run the initialization in a separate thread while tracking.
